File: android/src/main/java/com/cloudwebrtc/webrtc/model/IceCandidate.kt
package com.cloudwebrtc.webrtc.model

import org.webrtc.IceCandidate as WIceCandidate;

data class IceCandidate(val sdpMid: String, val sdpMLineIndex: Int, val sdp: String) {
    companion object {
        fun fromMap(map: Map<String, Any>): IceCandidate {
            return IceCandidate(
                map["sdpMid"] as String,
                map["sdpMLineIndex"] as Int,
                map["sdp"] as String
            )
        }

        fun fromWebRtc(from: WIceCandidate): IceCandidate =
            IceCandidate(from.sdpMid, from.sdpMLineIndex, from.sdp)

    }

    fun intoWebRtc(): WIceCandidate {
        return WIceCandidate(sdpMid, sdpMLineIndex, sdp)
    }

    fun intoMap(): Map<String, Any> = mapOf(
        "sdpMid" to sdpMid,
        "sdpMLineIndex" to sdpMLineIndex,
        "sdp" to sdp
    )
}

File: ios/Classes/model/IceCandidate.swift
import WebRTC

/// Representation of an `RTCIceCandidate`.
class IceCandidate {
  /// `mid` of this `IceCandidate`.
  var sdpMid: String

  /// `sdpMLineIndex` of this `IceCandidate`.
  var sdpMLineIndex: Int

  /// `candidate` of this `IceCandidate`.
  var candidate: String

  /// Creates a new `IceCandidate` object based on the method call received from the Flutter side.
  init(candidate: RTCIceCandidate) {
    self.sdpMid = candidate.sdpMid!
    self.candidate = candidate.sdp
    self.sdpMLineIndex = Int(candidate.sdpMLineIndex)

  }

  /// Creates new `IceCandidate` with a provided data.
  init(sdpMid: String, sdpMLineIndex: Int, candidate: String) {
    self.sdpMid = sdpMid
    self.sdpMLineIndex = sdpMLineIndex
    self.candidate = candidate
  }

  /// Converts this `IceCandidate` to an `RTCIceCandidate`.
  func intoWebRtc() -> RTCIceCandidate {
    RTCIceCandidate(
      sdp: self.candidate, sdpMLineIndex: Int32(self.sdpMLineIndex), sdpMid: self.sdpMid)
  }

  /// Converts this `IceCandidate` into a `Dictionary` which can be returned to the Flutter side.
  func asFlutterResult() -> [String: Any] {
    [
      "sdpMid": sdpMid,
      "sdpMLineIndex": sdpMLineIndex,
      "candidate": candidate,
    ]
  }
}
